### PR TITLE
Optimize managed FastTree Sumup to native parity on arm64

### PR DESCRIPTION
**Pull request by vladimir-aubrecht**
## Problem
FastTree builds feature histograms via `Sumup`, the per-iteration hot loop of tree
training. On x64/x86 this uses the native FastTree library; on **arm64** (and any
platform where the native library isn't available) it falls back to the generic
managed `IntArray.Sumup`, which goes through the `IIntArrayForwardIndexer` interface
with per-element bounds checks. That fallback is ~1.8x slower than native and
allocates on every call.

Note: the native `Sumup` (`Sumup.h`) is itself a plain **scalar** loop — the only real
SIMD in FastTreeNative is in `segment.cpp` (one-time segment compression), not in the
per-iteration histogram path. So there is no algorithmic reason managed can't match it.

## Change
Add optimized managed Sumup implementations that mirror the native templates
(`Sumup.h` / `SumupNibbles.h` / `SumupSegment.h`) exactly, using `fixed` pointers and
no bounds checks:

- **DenseIntArray**: new `SumupManagedDense` covering 4/8/16/32-bit × weighted/unweighted
  × root (no doc indices)/leaf. `Dense8/4/16/32BitIntArray` now point their managed
  handler at it instead of the slow `base.Sumup`.
- **SegmentIntArray**: new `SumupManaged` mirroring `SumupSegment` /
  `SumupSegment_noindices` for the compressed segment format.

Native remains the default on x64/x86 (`UseFastTreeNative` is unchanged) — only the
managed fallback path is replaced, so arm64 picks up the fast path automatically.
Because the loops iterate in the same order as native, float accumulation is
**bit-identical** and existing baselines are unchanged.

## Results (Apple M5, arm64)
Real in-repo types (`FeatureHistogram.SumupWeighted`), Dense8, weighted, 256 bins:

| case | N        | old fallback         | new managed           | speedup | identical |
|------|----------|----------------------|-----------------------|---------|-----------|
| root | 1,000,000| 1.119 ms (894 Melem/s)| 0.632 ms (1582 Melem/s)| 1.77x  | yes       |
| leaf | 500,000  | 0.553 ms (905 Melem/s)| 0.363 ms (1378 Melem/s)| 1.52x  | yes       |

Native comparison (native `Sumup.h` scalar loop built for arm64 vs equivalent managed,
N=20M): native 1337 Melem/s, old fallback 746 Melem/s (1.79x slower), new managed
1386 Melem/s (**0.96x = parity**). Managed allocations per call: 20 B -> **0 B**.

## Testing
- FastTree/FastForest/FastRank baseline tests in `Microsoft.ML.Predictor.Tests` pass on
  arm64 (7 passed, 0 failed; the managed path is what runs there, compared against
  native-generated baselines -> confirms numerical identity).
- Standalone verification of the two non-trivial decoders: segment bit-unpacking
  (2000 random trials, sequential + indexed) and the 4-bit nibble formula
  (5000 trials) — 0 mismatches.

## Notes for reviewers
- This does not change x64/x86 behavior; native stays the default there.
- `UseFastTreeNative` is intentionally left unchanged. If we later want to drop the
  native library entirely, this managed path is now fast enough to be the default on
  all architectures — but that is a separate decision.

Fixes #<issue-number>

### CODE REVIEW

**Comment by vladimir-aubrecht** [comment]
@dotnet-policy-service agree company="Microsoft"

File: src/Microsoft.ML.FastTree/Dataset/SegmentIntArray.cs [hunk 1]
```diff
@@ -576,6 +576,107 @@ public unsafe void SumupCPlusPlus(SumupInputData input, FeatureHistogram histogr
...
+        /// FastTree library is not available (e.g. arm64). This mirrors the native SumupSegment /
+        /// SumupSegment_noindices templates in src/Native/FastTreeNative/SumupSegment.h exactly,
+        /// including the segment bit-unpacking and accumulation order, so histogram results are
+        /// bit-identical to the native implementation. Reads go through fixed pointers to avoid the
+        /// per-element bounds checks and interface-indexer dispatch of the generic
+        /// <see cref="IntArray.Sumup"/> fallback.
+        /// </summary>
+        public unsafe void SumupManaged(SumupInputData input, FeatureHistogram histogram)
+        {
+            using (Timer.Time(TimerEvent.SumupSegment))
```
[the remark this answers is not in the document]
Comment (vladimir-aubrecht) [reply src/Microsoft.ML.FastTree/Dataset/SegmentIntArray.cs]: Fixed in bb334384d — removed the nested `Timer.Time(TimerEvent.SumupSegment)` from `SumupManaged`; timing is now done only by the public `Sumup` override.

**Comment by matouskozak** [comment]
> Standalone verification of the two non-trivial decoders: segment bit-unpacking
(2000 random trials, sequential + indexed) and the 4-bit nibble formula
(5000 trials) — 0 mismatches.

Would it make sense to add these as a tests? I'm wondering if we have much coverage over the new implementation...

I think ideally we should run both native and the new manage implementation on CI side by side to make sure they are ± equal.

**Comment by vladimir-aubrecht** [comment]
Good idea — added `FastTreeSumupParityTests` (946ffe675) that turns the ad-hoc verification into real tests:

- **`ManagedSumupMatchesReference`** — managed histogram vs an independent brute-force reference. Covers Dense 4/8/16/32-bit and Segment, root (no doc indices) and leaf (with doc indices), with and without weights. Runs on all platforms **including arm64**.
- **`ManagedSumupMatchesNative`** — the native and managed handlers run **side by side** on the same random data and the histograms are asserted **bit-identical**. It's gated with `[NativeDependencyTheory("FastTreeNative")]`, so it executes wherever the native FastTree library is present (i.e. the x64 CI legs) and is skipped where it isn't (arm64).

To let the tests call the managed path directly, the Dense `SumupManaged` handlers are now `internal` (Segment's is already public). I deliberately left `SumupNative` as-is to avoid an unrelated visibility change in this PR.

Note: the current red CI is an unrelated flaky `SdcaLogisticRegression` (LogLoss 0.5061 vs the 0.5 bound) on the arm64 Release leg — SDCA doesn't touch the FastTree Sumup path, the arm64 Debug leg passed, and it reproduces as borderline/non-deterministic locally.

**Comment by vladimir-aubrecht** [comment]
@matouskozak the parity tests are in and CI is green.

\`FastTreeSumupParityTests\` covers Dense 4/8/16/32-bit and Segment arrays, in both the root (no doc indices) and leaf (with doc indices) cases, with and without weights:

- **\`ManagedSumupMatchesReference\`** — the managed \`Sumup\` histogram is checked against an independent brute-force reference. Runs on every platform, including arm64.
- **\`ManagedSumupMatchesNative\`** — native and managed run **side by side** on the same data and the histograms are asserted **bit-identical**; both are also checked against the reference (so a shared decode mistake cannot hide behind an equal-but-wrong comparison). Gated with \`[NativeDependencyTheory("FastTreeNative")]\`, so it executes wherever the native library is present (the x64 CI legs) and is skipped elsewhere.

Two things worth calling out that surfaced while wiring this up:

1. Building a \`SegmentIntArray\` through \`IntArray.New(..., Segmented, ...)\` runs the native \`C_SegmentFindOptimalPath\`, which types its buffer as \`unsigned long*\` — 64-bit on LP64 (Linux/macOS x64) while the managed array is 32-bit \`uint[]\`, so it overruns the buffer (SIGSEGV). It is pre-existing and unrelated to the Sumup decoders here; the test builds the segment array with the managed encoder to avoid it. Happy to file a separate issue for the native encoder.
2. Segment arrays built via \`FromWorkArray\` do not have \`SumupHandler\` wired up, so the test calls the native decoder (\`SumupCPlusPlus\`) directly rather than \`arr.Sumup()\`.

The only change to product code is making the Dense \`SumupManaged\` handlers \`internal\` so the tests can call them directly; \`SumupNative\` is left as-is.

File: test/Microsoft.ML.Tests/FastTreeSumupParityTests.cs [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+    /// Verifies the optimized managed <c>Sumup</c> implementations added for platforms without the
+    /// native FastTree library (e.g. arm64). Two properties are checked for Dense 4/8/16/32-bit and
+    /// Segment arrays, in both the root (no doc indices) and leaf (with doc indices) cases:
+    /// <list type="bullet">
+    /// <item>the managed histogram matches an independent brute-force reference (runs everywhere), and</item>
+    /// <item>the managed histogram is bit-identical to the native histogram (runs where FastTreeNative exists,
+    /// i.e. x64 CI legs), giving the "native and managed side by side" coverage requested in the PR review.</item>
+    /// </list>
+    /// </summary>
+    public sealed class FastTreeSumupParityTests : BaseTestClass
```
Comment (matouskozak) [inline test/Microsoft.ML.Tests/FastTreeSumupParityTests.cs:27]: I might have missed it but I don't see much corner case tests and tests that produce multiple segments of different lengths (which is the difficult part). I'm not sure if the existing native implementation also tests this or not but if it does, we should probably test it also on the managed side to be sure.

File: src/Microsoft.ML.FastTree/Dataset/DenseIntArray.cs [hunk 3]
```diff
@@ -457,6 +574,17 @@ private void SumupNative(SumupInputData input, FeatureHistogram histogram)
             }
         }
 
+        internal void SumupManaged(SumupInputData input, FeatureHistogram histogram)
```
Comment (matouskozak) [inline src/Microsoft.ML.FastTree/Dataset/DenseIntArray.cs:577]: Would it make sense to add a test that would validated that on arm64 we're really going through the managed path and x64 remains on native?

**Review by matouskozak — APPROVED** [review]
Overall LGTM!. I left a few comments related to the test coverage.

File: test/Microsoft.ML.Tests/FastTreeSumupParityTests.cs (hunk 2 above)
Comment (vladimir-aubrecht) [reply test/Microsoft.ML.Tests/FastTreeSumupParityTests.cs:27]: Added `ManagedSegmentSumupHandlesVariedShapes` (430062979). It builds segment arrays whose value magnitudes vary in runs, producing **multiple segments of different bit widths and run lengths**, and covers the tricky cases explicitly: transitions across 32-bit packed-word boundaries (`WordBoundaryWidths`, `PowersOfTwo`), long single-width runs spanning many words (`LongSingleWidth`), `IncreasingWidths`/`DecreasingWidths`/`ManyShortSegments`/`AlternatingExtremes`, plus degenerate corner cases (`AllZero`, `AllSame`, `Single`, `Two`) — each with/without weights and root/leaf. Managed is compared to an independent reference, and where FastTreeNative is present (x64) also asserted bit-identical to native, so the same multi-segment shapes are validated on both sides.

File: src/Microsoft.ML.FastTree/Dataset/DenseIntArray.cs (hunk 3 above)
Comment (vladimir-aubrecht) [reply src/Microsoft.ML.FastTree/Dataset/DenseIntArray.cs:577]: Added `SumupDispatchMatchesArchitecture` (430062979). It asserts `IntArray.UseFastTreeNative` matches the process architecture (true on x64/x86, false otherwise) and, to validate the actual wiring rather than just the flag, reads back the `SumupHandler` delegate on a real array via reflection and checks it resolves to `SumupNative` on x64/x86 and `SumupManaged` on arm64. Verified locally on arm64 that it resolves to the managed handler.

**Comment by vladimir-aubrecht** [comment]
Thanks for the review! Addressed both test-coverage comments in 430062979: added multi-segment/corner-case coverage for the segment decoder and a test asserting the arm64→managed / x64→native dispatch. Let me know if you'd like anything else.

**Review by matouskozak — APPROVED** [review]
